src/sink/rest.rs

```diff
@@ -837,30 +842,69 @@ async fn build_response_events(
     event_id: &EventId,
     response_ids: &mut EventIdGenerator,
     event_origin_uri: &EventOriginUri,
+    request: Request,
     mut response: Response,
     codec: &mut dyn Codec,
     codec_map: &mut HashMap<String, Box<dyn Codec>>,
     preprocessors: &mut [Box<dyn Preprocessor>],
 ) -> Result<Vec<Event>> {
     let mut meta = Value::object_with_capacity(1);
     let mut response_meta = Value::object_with_capacity(2);
+    let mut request_meta = Value::object_with_capacity(3);
 
     let numeric_status: u16 = response.status().into();
     response_meta.insert("status", numeric_status)?;
 
-    let mut headers = Value::object_with_capacity(8);
+    let mut response_headers = Value::object_with_capacity(8);
     {
         for (name, values) in response.iter() {
             let header_values: Value = values
                 .iter()
                 .map(ToString::to_string)
                 .map(Value::from)
                 .collect();
-            headers.insert(name.to_string(), header_values)?;
+            response_headers.insert(name.to_string(), header_values)?;
         }
     }
-    response_meta.insert("headers", headers)?;
+
+    response_meta.insert("headers", response_headers)?;
+    request.content_type();
+
+    let method = request.method().to_string();
+    request_meta.insert("method", method)?;
+
+    let mut request_headers = Value::object_with_capacity(request.header_values().count());
+    {
+        for (name, values) in request.iter() {
+            let header_values: Value = values
+                .iter()
+                .map(ToString::to_string)
+                .map(Value::from)
+                .collect();
+            request_headers.insert(name.to_string(), header_values)?;
+        }
+    }
+    request_meta.insert("headers", request_headers)?;
+
+    let mut endpoint = Value::object_with_capacity(8);
+    endpoint.insert("scheme", none_if_empty(request.url().scheme()))?;
+    endpoint.insert("username", none_if_empty(request.url().username()))?;
+    endpoint.insert(
+        "password",
+        request.url().password().map(ToString::to_string),
+    )?;
+    endpoint.insert("host", request.url().host_str().map(ToString::to_string))?;
+    endpoint.insert("port", request.url().port())?;
+    endpoint.insert("path", none_if_empty(request.url().path()))?;
+    endpoint.insert("query", request.url().query().map(ToString::to_string))?;
+    endpoint.insert(
+        "fragment",
+        request.url().fragment().map(ToString::to_string),
+    )?;
+    request_meta.insert("endpoint", endpoint)?;
+
     meta.insert("response", response_meta)?;
+    meta.insert("request", request_meta)?;
```

Comment:
can we put this underneath the `response` meta, so this lives at `$response.request`?

The reason is that the Rest offramp will interpret the `$request` metadata and creates a HTTP request from it, so depending on how people wire offramps off, it is possible that this request metadata might interfere with what the user has to configure for the Rest offramp. So we want to avoid this interference.